Foundation/src/Process_UNIX.cpp

```diff
@@ -182,76 +182,83 @@ ProcessHandleImpl* ProcessImpl::launchByForkExecImpl(const std::string& command,
+		// therefore allocate all required buffers first.
 
-	int pid = fork();
-	if (pid < 0)
-	{
-		throw SystemException("Cannot fork process for", command);
-	}
-	else if (pid == 0)
-	{
-		if (pInitialDirectory)
+		std::vector<char> envChars = getEnvironmentVariablesBuffer(env);
+		std::vector<char*> argv(args.size() + 2);
+		int i = 0;
+		argv[i++] = const_cast<char*>(command.c_str());
+		for (const auto& a: args)
 		{
-			if (chdir(pInitialDirectory) != 0)
-			{
-				_exit(72);
-			}
+			argv[i++] = const_cast<char*>(a.c_str());
 		}
+		argv[i] = NULL;
 
-		// set environment variables
-		char* p = &envChars[0];
-		while (*p)
-		{
-			putenv(p);
-			while (*p) ++p;
-			++p;
-		}
+		const char* pInitialDirectory = initialDirectory.empty() ? 0 : initialDirectory.c_str();
 
-		// setup redirection
-		if (inPipe)
+		int pid = fork();
+		if (pid < 0)
 		{
-			dup2(inPipe->readHandle(), STDIN_FILENO);
-			inPipe->close(Pipe::CLOSE_BOTH);
+			throw SystemException("Cannot fork process for", command);
 		}
-		if (options & PROCESS_CLOSE_STDIN) close(STDIN_FILENO);
-
-		// outPipe and errPipe may be the same, so we dup first and close later
-		if (outPipe) dup2(outPipe->writeHandle(), STDOUT_FILENO);
-		if (errPipe) dup2(errPipe->writeHandle(), STDERR_FILENO);
-		if (outPipe) outPipe->close(Pipe::CLOSE_BOTH);
-		if (options & PROCESS_CLOSE_STDOUT) close(STDOUT_FILENO);
-		if (errPipe) errPipe->close(Pipe::CLOSE_BOTH);
-		if (options & PROCESS_CLOSE_STDERR) close(STDERR_FILENO);
-		// close all open file descriptors other than stdin, stdout, stderr
-		long fdMax = sysconf(_SC_OPEN_MAX);
-		// on some systems, sysconf(_SC_OPEN_MAX) returns a ridiculously high number
-		if (fdMax > CLOSE_FD_MAX) fdMax = CLOSE_FD_MAX;
-		for (long i = 3; i < fdMax; ++i)
+		else if (pid == 0)
 		{
-			close(i);
+			if (pInitialDirectory)
+			{
+				if (chdir(pInitialDirectory) != 0)
+				{
+					break;
+				}
+			}
+
+			// set environment variables
+			char* p = &envChars[0];
+			while (*p)
+			{
+				putenv(p);
+				while (*p) ++p;
+				++p;
+			}
+
+			// setup redirection
+			if (inPipe)
+			{
+				dup2(inPipe->readHandle(), STDIN_FILENO);
+				inPipe->close(Pipe::CLOSE_BOTH);
+			}
+			if (options & PROCESS_CLOSE_STDIN) close(STDIN_FILENO);
+
+			// outPipe and errPipe may be the same, so we dup first and close later
+			if (outPipe) dup2(outPipe->writeHandle(), STDOUT_FILENO);
+			if (errPipe) dup2(errPipe->writeHandle(), STDERR_FILENO);
+			if (outPipe) outPipe->close(Pipe::CLOSE_BOTH);
+			if (options & PROCESS_CLOSE_STDOUT) close(STDOUT_FILENO);
+			if (errPipe) errPipe->close(Pipe::CLOSE_BOTH);
+			if (options & PROCESS_CLOSE_STDERR) close(STDERR_FILENO);
+			// close all open file descriptors other than stdin, stdout, stderr
+			long fdMax = sysconf(_SC_OPEN_MAX);
+			// on some systems, sysconf(_SC_OPEN_MAX) returns a ridiculously high number
+			if (fdMax > CLOSE_FD_MAX) fdMax = CLOSE_FD_MAX;
+			for (long i = 3; i < fdMax; ++i)
```

Comment:
@pavledragisic let's also fix this CodeQL alert

Reply:
Fixed in [8c4d83c](https://github.com/pocoproject/poco/pull/4267/commits/8c4d83c86860d1730a71d29b58aae1e9d8080db0)